pulsar-io/elastic-search/src/main/java/org/apache/pulsar/io/elasticsearch/ElasticSearchSink.java

```diff
@@ -18,139 +18,254 @@
+            keySchema = keyValueSchema.getKeySchema();
+            valueSchema = keyValueSchema.getValueSchema();
+            if (KeyValueEncodingType.SEPARATED.equals(keyValueSchema.getKeyValueEncodingType()) && record.getKey().isPresent()) {
+                key = keySchema.decode(record.getKey().get().getBytes(StandardCharsets.UTF_8));
+            }
+            if (record.getValue() != null) {
+                value = ((KeyValue) record.getValue().getNativeObject()).getValue();
+                if (KeyValueEncodingType.INLINE.equals(keyValueSchema.getKeyValueEncodingType())) {
+                    key = ((KeyValue) record.getValue().getNativeObject()).getKey();
+                }
+            }
+        } else {
+            key = record.getKey().orElse(null);
+            valueSchema = record.getSchema();
+            value = record.getValue() == null ? null : record.getValue().getNativeObject();
+        }
 
-        if (!exists) {
-            CreateIndexRequest cireq = new CreateIndexRequest(elasticSearchConfig.getIndexName());
+        String id = key.toString();
+        if (keySchema != null) {
+            id = stringifyKey(keySchema, key);
+        }
 
-            cireq.settings(Settings.builder()
-               .put("index.number_of_shards", elasticSearchConfig.getIndexNumberOfShards())
-               .put("index.number_of_replicas", elasticSearchConfig.getIndexNumberOfReplicas()));
+        String doc = null;
+        if (value != null) {
+            if (valueSchema != null) {
+                doc = stringifyValue(valueSchema, value);
+            } else {
+                doc = value.toString();
+            }
+        }
 
-            CreateIndexResponse ciresp = getClient().indices().create(cireq, RequestOptions.DEFAULT);
-            if (!ciresp.isAcknowledged() || !ciresp.isShardsAcknowledged()) {
-                throw new RuntimeException("Unable to create index.");
+        if (elasticSearchConfig.isKeyIgnore()) {
+            if (doc == null || Strings.isNullOrEmpty(elasticSearchConfig.getPrimaryFields())) {
+                id = Hex.encodeHexString(record.getMessage().get().getMessageId().toByteArray());
+            } else {
+                try {
+                    // extract the PK from the JSON document
+                    JsonNode jsonNode = objectMapper.readTree(doc);
+                    List<String> pkFields = Arrays.asList(elasticSearchConfig.getPrimaryFields().split(","));
+                    id = stringifyKey(jsonNode, pkFields);
+                } catch (JsonProcessingException e) {
+                    log.error("Failed to read JSON", e);
+                }
             }
         }
+
+        SchemaType schemaType = null;
+        if (record.getSchema() != null && record.getSchema().getSchemaInfo() != null) {
+            schemaType = record.getSchema().getSchemaInfo().getType();
+        }
+        log.debug("recordType={} schemaType={} id={} doc={}",
```

Comment:
please add:
`if (log.isDebugEnabled()....)`

Reply:
You mean add if(log.isDebugEnabled()) before every log.debug() ? The log.debug() is doing that check internally, this is useless ?

Reply:
yes. I mean to add it.
it is a pattern in Pulsar code base and we should keep the code consistent.

there is a never ending debate about this topic :-) 

my take is that if a project uses one style it is better to follow that codestyle.

if we want to change it is better to open a separate discussion

Reply:
Awesome !

Reply:
It's already in a` if (log.isDebugEnabled()) {} ` block here, that's not enough ?